Add event logging with application insights

LoganHollins opened the pull request:
Related: 
https://github.com/bcjobs/jobcentre-com/pull/218
https://github.com/bcjobs/jobcentre/pull/284

LoganHollins committed:
Add event logging with application insights

LoganHollins committed:
Change sessionStorage -> localStorage; Allow jobId to be int or string

LoganHollins committed:
Replace old click with new click

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,55 @@
+      if(clickArray == null)
+          clickArray = [];
+
+      var clickIndex = clickArray.map(function(e){return e.jobId;}).indexOf(parseInt(jobId));
```
We haven't polyfilled `.map`, so will break in older browsers.  Better to use underscore's `_.map()` (http://underscorejs.org/#map) which is loaded on every page.

johnnyoshika replied:
We should always pass in the `radix` to `parseInt`:

```
parseInt(jobId, 10)
```

More info here: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/parseInt

![image](https://user-images.githubusercontent.com/504505/29797807-287d0c98-8c0e-11e7-80c3-eb8f7d6f7071.png)

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,55 @@
+          return;
+
+      // Only find index if jobEvent hasn't happened previously
+      var clickIndex = clickArray.map(function(e){return e.jobId;}).indexOf(parseInt(jobId));
```
Same comment as above re: map and parseInt.

johnnyoshika commented:
@LoganHollins Can you please do me one more favour? Can you test this in IE8? Our site doesn't need to look good in IE8, but I still want people to be able to use it.  Using IE11 in IE8 mode is good enough. Thanks!

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,55 @@
+
+      var clickArray = JSON.parse(window.localStorage.getItem('recommendationInfo'));
+      if(clickArray == null)
+          clickArray = [];
```
Instead of storing an array like this:

```
[{
    "jobId": 269891,
    "recommender": "MoreLikeThis",
    "timestamp": "2017-08-28T23:53:50.611Z"
}, {
    "jobId": 269716,
    "recommender": "MoreLikeThis",
    "timestamp": "2017-08-28T23:54:20.569Z"
}]
```

I suggest you use an object which acts as a hash table.  Finding and replacing would be much easier (and faster) this way:

```
{
    "269891": {
      "recommender": "MoreLikeThis",
      "timestamp": "2017-08-28T23:53:50.611Z"
  },
  "269716": {
      "recommender": "MoreLikeThis",
      "timestamp": "2017-08-28T23:54:20.569Z"
  }
}
```

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,55 @@
+      if(!JOBCENTRE.capabilities.localStorage)
+          return;
+
+      var click = {'jobId': parseInt(jobId), 'recommender': recommender, 'timestamp': new Date().toISOString()};
```
I think it would be easier to work with epoch (`new Date().getTime()`) instead of the string representation.  That way, you don't have to parse the string to do comparisons.

johnnyoshika commented:
One unfortunate thing about fetching data from localStorage is that you can never really trust the data there.  You can't trust that it's proper JSON, and even if it is, you can't trust that it has the properties you expect.  Things get modified.  I'm not sure how (whether a user tinkers with it, an extension tinkers with it, so something get corrupted), but malformed data somehow seems to show up in production.  So let's protect against it.  Here's an example of how I do it in job_form.js as an example:

https://github.com/bcjobs/assets/blob/master/js/debug/job_form.js#L264-L281

The idea to make sure the script continues and recovers well regardless of what data is in localStorage.

LoganHollins committed:
Use hash table instead of array for storing clicks

LoganHollins commented:
@johnnyoshika I have updated the PR using a hash table to store the clicks.  I also updated `com` and made it pass the Job Id as an int to the view method.

LoganHollins commented:
@johnnyoshika I tested IE8 using IE11 in IE8 mode, and it didn't cause any problems.  The pages loaded and worked properly, even when submitting an application.

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,53 @@
+            clicks = {};
+
+        clicks[jobId] = {'recommender': recommender, 'timestamp': new Date().getTime()};
+        localStorage.setItem('recommendationInfo', JSON.stringify(clicks));
```
How about we move all of this stuff up inside the try catch? That way, even if `clicks` is not an object (i.e. it's a string or an array), this code will fail gracefully.

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,53 @@
+
+        clicks[jobId] = click;
+        localStorage.setItem('recommendationInfo', JSON.stringify(clicks));
+        appInsights.trackEvent('JobseekerRecommendedJob' + jobEvent, {'jobId': jobId, 'recommender': click.recommender});
```
Let's move all of this stuff up into try/catch.  That way, even if `click.timestamp` isn't a number, we won't get an exception and everything will fail gracefully.  We could then also simplify line 41 as we won't have to do so many checks.

LoganHollins committed:
Move all code into try catch block; make jobId a string

LoganHollins commented:
@johnnyoshika I have updated this PR and moved everything inside the try catch block. I have also updated jobId to be a string.

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,54 @@
+        if(!JOBCENTRE.capabilities.localStorage || !localStorage.getItem('recommendationInfo'))
+            return;
+
+        var clicks;
```
This is no longer necessary as `clicks` is all inside the try/catch.

johnnyoshika commented on js/debug/job_recommendation.js:
```diff
@@ -0,0 +1,54 @@
+        var clicks;
+        try {
+            jobId += '';
+            clicks = JSON.parse(localStorage.getItem('recommendationInfo'));
```
Let's scope this with `var` once we remove the `clicks` declaration above.

johnnyoshika reviewed (changes requested):
A couple of minor fixes and I think this will be it!

LoganHollins committed:
Move clicks declaration inside try catch

LoganHollins commented:
@johnnyoshika The PR has been updated and `clicks` is now declared inside the try catch blocks.
